src/backend/oneapi/convolve.cpp

```diff
@@ -110,18 +110,54 @@ template<typename T>
 Array<T> convolve2_unwrap(const Array<T> &signal, const Array<T> &filter,
                           const dim4 &stride, const dim4 &padding,
                           const dim4 &dilation) {
-    Array<T> out =
-        convolve2_unwrap<T>(signal, filter, stride, padding, dilation);
+    dim4 sDims = signal.dims();
+    dim4 fDims = filter.dims();
+
+    dim_t outputWidth =
+        1 + (sDims[0] + 2 * padding[0] - (((fDims[0] - 1) * dilation[0]) + 1)) /
+                stride[0];
+    dim_t outputHeight =
+        1 + (sDims[1] + 2 * padding[1] - (((fDims[1] - 1) * dilation[1]) + 1)) /
+                stride[1];
+
+    const bool retCols = false;
+    Array<T> unwrapped =
+        unwrap(signal, fDims[0], fDims[1], stride[0], stride[1], padding[0],
+               padding[1], dilation[0], dilation[1], retCols);
+
+    unwrapped  = reorder(unwrapped, dim4(1, 2, 0, 3));
+    dim4 uDims = unwrapped.dims();
+
+    unwrapped =
+        modDims(unwrapped, dim4(uDims[0] * uDims[1], uDims[2] * uDims[3]));
+
+    Array<T> collapsedFilter = filter;
+
+    collapsedFilter = flip(collapsedFilter, {1, 1, 0, 0});
+    collapsedFilter = modDims(collapsedFilter,
+                              dim4(fDims[0] * fDims[1] * fDims[2], fDims[3]));
+
+    Array<T> res =
+        matmul(unwrapped, collapsedFilter, AF_MAT_TRANS, AF_MAT_NONE);
+    res = modDims(res, dim4(outputWidth, outputHeight, signal.dims()[3],
+                            collapsedFilter.dims()[1]));
+    Array<T> out = reorder(res, dim4(0, 1, 3, 2));
 
     return out;
 }
 
 template<typename T>
 Array<T> convolve2(Array<T> const &signal, Array<T> const &filter,
                    const dim4 stride, const dim4 padding, const dim4 dilation) {
-    ONEAPI_NOT_SUPPORTED("");
-    Array<T> out = createEmptyArray<T>(dim4(1));
-    return out;
+    if constexpr (!std::is_same<T, half>::value) {
```

Comment:
if this compiles with half enabled (dropping the if constexpr), the compiler complains with:
```
/home/gpryor/arrayfire/src/backend/oneapi/blas.hpp:33:24: error: no viable conversion from 'double' to 'const arrayfire::common::half'
    static constexpr T alpha = 1.0;
```
casting or `1.0f` has no effect on the error message. i vote we move on from half from here on out since this seems like a deeper issue